src/Core/Manifest/PrioritySorter.php

```diff
@@ -0,0 +1,216 @@
+<?php
+
+namespace SilverStripe\Core\Manifest;
+
+use SilverStripe\Core\Injector\Injectable;
+
+/**
+ * Class PrioritySorter
+ * @package SilverStripe\Core\Manifest
+ */
+class PrioritySorter
+{
+    use Injectable;
+
+    /**
+     * The key that is used to denote all remaining items that have not
+     * been specified in priorities
+     * @var string
+     */
+    protected $restKey = '...rest';
+
+    /**
+     * A map of variables to their values
+     * @var array
+     */
+    protected $variables = [];
+
+    /**
+     * @var array
+     */
+    protected $items;
+
+    /**
+     * @var array
+     */
+    protected $priorities;
+
+    /**
+     * @var array
+     */
+    protected $names;
+
+    /**
+     * If not specified in priorities, put this item at the top (can be a variable)
+     *
+     * @var string
+     */
+    protected $defaultTop;
+
+    /**
+     * PrioritySorter constructor.
+     * @param array $items
+     * @param array $priorities
+     */
+    public function __construct(array $items = [], array $priorities = [])
+    {
+        $this->setItems($items);
+        $this->priorities = $priorities;
+    }
+
+    /**
+     * Sorts the items and returns a new version of $this->items
+     *
+     * @return array
+     */
+    public function getSortedList()
+    {
+        $this->addVariables();
+
+        // Find all items that don't have their order specified by the config system
+        $unspecified = array_diff($this->names, $this->priorities);
+
+        if ($this->restKey && !empty($unspecified)) {
+            $this->includeRest($unspecified);
+        }
+
+        if ($this->defaultTop) {
+            $this->includeDefaultTop();
+        }
+
+        $sortedList = [];
+        foreach ($this->priorities as $itemName) {
+            if (isset($this->items[$itemName])) {
+                $sortedList[$itemName] = $this->items[$itemName];
+            }
+        }
+        $sortedList = array_reverse($sortedList, true);
```

Comment:
I think we are double-reversing this list. If you put `$project` earlier in the config, you don't need to reverse the array here.

Also my testing is showing your $sortedList as 

```

[0] =>
--
  | string(22) "silverstripe/versioned"
  | [1] =>
  | string(23) "silverstripe/siteconfig"
  | [2] =>
  | string(22) "silverstripe/errorpage"
  | [3] =>
  | string(20) "silverstripe/reports"
  | [4] =>
  | string(19) "silverstripe/assets"
  | [5] =>
  | string(18) "silverstripe/admin"
  | [6] =>
  | string(6) "mysite"
  | [7] =>
  | string(20) "silverstripe/graphql"
  | [8] =>
  | string(22) "silverstripe/framework"
  | [9] =>
  | string(16) "silverstripe/cms"
  | [10] =>
  | string(27) "silverstripe/campaign-admin"
  | [11] =>
  | string(24) "silverstripe/asset-admin"

```

Not sure why mysite is ending up in the middle.